### PR TITLE
MachineProcessCompleteEvent

### PR DESCRIPTION
**Pull request by poma123**
## Description

Added MachineProcessCompleteEvent, which fires whenever an AContainer has completed its process. This was the best name I could came up with. If you have a better idea, then tell me.

I can add a Unit Test, if needed.

## Changes

- Added MachineProcessCompleteEvent.class
- Fixed a small typo in GEOResourceGenerationEvent.class
- Added the #callEvent() method to AContainer.class
- Finally a minor impovement in AContainer.class

## Related Issues

N/A

## Checklist

- [x] I have fully tested the proposed changes and promise that they will not break everything into chaos.
- [x] I have also tested the proposed changes in combination with various popular addons and can confirm my changes do not break them.
- [x] I followed the existing code standards and didn't mess up the formatting.
- [x] I did my best to add documentation to any public classes or methods I added.
- [x] I have added `Nonnull` and `Nullable` annotations to my methods to indicate their behaviour for null values
- [ ] I added sufficient Unit Tests to cover my code.

### CODE REVIEW

**Comment by TheBusyBiscuit** [comment]
Oh hmmm, I am not too sure whether the Event API would be the most suitable for this. To be honest the entire logic behind machine processing is long overdue for a rewrite anyway and we could probably implement this in a much better way, as firing this many events can create a bit of a performance disadvantage.

File: src/main/java/io/github/thebusybiscuit/slimefun4/api/events/MachineProcessCompleteEvent.java [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+ *
+ */
+public class MachineProcessCompleteEvent extends Event {
+
+    private static final HandlerList handlerList = new HandlerList();
+
+    private final Block block;
+    private final MachineRecipe machineRecipe;
+
+    @ParametersAreNonnullByDefault
```
Comment (WalshyDev) [inline src/main/java/io/github/thebusybiscuit/slimefun4/api/events/MachineProcessCompleteEvent.java:25]: You already tag them as Nonnull...

**Review by WalshyDev — CHANGES_REQUESTED** [review]

File: src/main/java/io/github/thebusybiscuit/slimefun4/api/events/MachineProcessCompleteEvent.java (hunk 1 above)
Comment (poma123) [reply src/main/java/io/github/thebusybiscuit/slimefun4/api/events/MachineProcessCompleteEvent.java:25]: Reee

Comment (TheBusyBiscuit) [reply src/main/java/io/github/thebusybiscuit/slimefun4/api/events/MachineProcessCompleteEvent.java:25]: Reeeeee

**Comment by LinoxGH** [comment]
> Oh hmmm, I am not too sure whether the Event API would be the most suitable for this. To be honest the entire logic behind machine processing is long overdue for a rewrite anyway and we could probably implement this in a much better way, as firing this many events can create a bit of a performance disadvantage.

Agreed with the rewrite. But event api wouldn't be a problem, I don't think. Bukkit runs hundreds of events consistently without a problem.

**Comment by TheBusyBiscuit** [comment]
> Bukkit runs hundreds of events consistently without a problem.

Just because they are doing so doesn't mean it is a good idea :LUL:
Even the creation of new objects, like the event objects is already a drawback in terms of memory/performance that doesn't **need** to happen. It won't make such an enormous difference but the problem certainly scales and Slimefun is already pretty packed with features and events.

So I really think a rewrite (with such possibility to tap into the outcome) would be the best approach here.
I may probably still merge it for the time being though until that rewrite has happened.

**Review by TheBusyBiscuit — APPROVED** [review]

**Pull request merged**